File: schemas/messaging-chat/drizzle/message_reports.ts
// message_reports: User-submitted reports for message moderation.
// See README.md for full design rationale.
import { pgTable, pgEnum, uuid, text, timestamp, index, unique } from "drizzle-orm/pg-core";
import { messages } from "./messages";
import { users } from "./users";

export const messageReportReasonEnum = pgEnum("message_report_reason", [
  "spam",
  "harassment",
  "hate_speech",
  "violence",
  "misinformation",
  "nsfw",
  "other",
]);

export const messageReportStatusEnum = pgEnum("message_report_status", [
  "pending",
  "reviewed",
  "resolved",
  "dismissed",
]);

export const messageReports = pgTable(
  "message_reports",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    messageId: uuid("message_id")
      .notNull()
      .references(() => messages.id, { onDelete: "cascade" }),
    reporterId: uuid("reporter_id")
      .notNull()
      .references(() => users.id, { onDelete: "cascade" }),
    reason: messageReportReasonEnum("reason").notNull(),
    description: text("description"),
    status: messageReportStatusEnum("status").notNull().default("pending"),
    reviewedBy: uuid("reviewed_by").references(() => users.id, {
      onDelete: "set null",
    }),
    reviewedAt: timestamp("reviewed_at", { withTimezone: true }),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (t) => [
    unique().on(t.messageId, t.reporterId),
    index().on(t.status),
    index().on(t.reporterId),
    index().on(t.reviewedBy),
  ]
);

File: schemas/messaging-chat/spacetimedb/message_reports.rs
// message_reports: tracks user reports on messages for moderation review.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

#[derive(SpacetimeType, Clone)]
pub enum MessageReportReason {
    Spam, // type: String
    Harassment,
    HateSpeech,
    Violence,
    Misinformation,
    Nsfw,
    Other,
}

#[derive(SpacetimeType, Clone)]
pub enum MessageReportStatus {
    Pending, // type: String
    Reviewed,
    Resolved,
    Dismissed,
}

#[spacetimedb::table(name = message_reports, public)]
pub struct MessageReport {
    #[primary_key]
    pub id: String, // UUID

    // FK → messages.id (cascade delete)
    #[index(btree)]
    pub message_id: String, // UUID

    // FK → users.id (cascade delete)
    #[index(btree)]
    pub reporter_id: String, // UUID

    pub reason: MessageReportReason,

    pub description: Option<String>,

    // default: MessageReportStatus::Pending
    #[index(btree)]
    pub status: MessageReportStatus,

    // FK → users.id (set null on delete)
    #[index(btree)]
    pub reviewed_by: Option<String>, // UUID

    pub reviewed_at: Option<Timestamp>, // microseconds since epoch (nullable)

    pub created_at: Timestamp, // microseconds since epoch

    pub updated_at: Timestamp, // microseconds since epoch, updated on change

    // COMPOSITE UNIQUE: (message_id, reporter_id) — enforced in application logic
}
